src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb

```diff
@@ -303,6 +301,147 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+
+                Dim boundArgument As BoundExpression = BindRValue(argumentSyntax.Expression, diagnostics)
+                boundArguments.Add(boundArgument)
+
+                Dim elementType = GetTupleFieldType(boundArgument, argumentSyntax, diagnostics, hasErrors)
+                elementTypes.Add(elementType)
+
+                If elementType Is Nothing Then
+                    hasNaturalType = False
+                End If
+            Next
+
+            uniqueFieldNames.Free()
+
+            If countOfExplicitNames <> 0 AndAlso countOfExplicitNames <> elementTypes.Count Then
+                hasErrors = True
+                ReportDiagnostic(diagnostics, node, ERRID.ERR_TupleExplicitNamesOnAllMembersOrNone)
+            End If
+
+            Dim elementNamesArray = If(elementNames Is Nothing, Nothing, elementNames.ToImmutableAndFree())
+
+            Dim tupleTypeOpt As NamedTypeSymbol = Nothing
+            Dim elements = elementTypes.ToImmutableAndFree()
+            Dim locations = elementLocations.ToImmutableAndFree()
+
+            If hasNaturalType Then
+                tupleTypeOpt = TupleTypeSymbol.Create(node.GetLocation, elements, locations, elementNamesArray, Me.Compilation, node, diagnostics)
+            Else
+                TupleTypeSymbol.VerifyTupleTypePresent(elements.Length, node, Me.Compilation, diagnostics)
+            End If
+
+            Return New BoundTupleLiteral(node, elementNamesArray, boundArguments.ToImmutableAndFree(), tupleTypeOpt, hasErrors)
+        End Function
+
+        ''' <summary>
+        ''' Returns the type to be used as a field type generates errors in case the type is not
+        ''' supported for tuple type fields.
+        ''' </summary>
+        Private Function GetTupleFieldType(expression As BoundExpression, errorSyntax As VisualBasicSyntaxNode, diagnostics As DiagnosticBag, <Out> ByRef hasError As Boolean) As TypeSymbol
+            Dim expressionType As TypeSymbol = expression.Type
+
+            If Not expression.HasErrors Then
+                If expression.Type IsNot Nothing Then
+                    If expressionType.SpecialType = SpecialType.System_Void Then
+                        expressionType = ErrorTypeSymbol.UnknownResultType
+                        hasError = True
+                        ReportDiagnostic(diagnostics, errorSyntax, ERRID.ERR_VoidValue)
+
+                    ElseIf (expressionType.IsRestrictedType()) Then
+                        hasError = True
+                        ReportDiagnostic(diagnostics, errorSyntax, ERRID.ERR_RestrictedType1, expressionType)
+
+                    End If
+                End If
+            End If
+
+            Return expressionType
+        End Function
+
+        Private Shared Sub CollectTupleFieldMemberNames(name As String, position As Integer, tupleSize As Integer, ByRef elementNames As ArrayBuilder(Of String))
+            ' add the name to the list
+            ' names would typically all be there Or none at all
```

Comment:
Fixed

---
In reply to: [71441608](https://github.com/dotnet/roslyn/pull/12596#discussion_r71441608) [](ancestors = 71441608)